### ISSUE TITLE
custom: add cpuid and kconfig rules

### ISSUE THREAD

**Issue by mythi** [kubernetes-sigs/node-feature-discovery#321]
I have use-cases to build custom labels from cpuid, kconfig, and iommu and I started looking how to add these to the custom source.

The initial custom source implementation moved usb/pci utils to `source/internal` to share code between the custom source and pci/usb sources.

I opened this issue to discuss what'd be the right approach to start adding more rules to custom source.

- start moving common code to `source/internal`?
- re-think custom source and have, e.g,  a "mutating" whitelist filter?

the worker config has source specific fields. If we start moving the code we'd need to think how to deal with, e.g.,

```
#  kernel:
#    kconfigFile: "/path/to/kconfig"
```
for code that is not part of any source.

**Comment by marquiz** [comment]
For some time now, and especially after #290, I've been thinking that we should separate the discovery and label generation functions. We would have a detection layer that would do the system discovery and present the information via a unified and cached API/structure. Feature labels would then be created by filtering/combining this data. This would provide re-usability of the data (and diminishing the risk for races or incoherent view of the system) avoiding duplicate discovery passes where feature labels are based on the same information (e.g. `pci` and `usb` vs. `custom` source, currently). Caching would also allow us to disable re-discovery for information that is known not to change (and where detection is possibly time-consuming and costly). This change would be a bit bigger re-structuring, though, and takes some effort

I think for the short term the approach would be to implement new matchers to the `custom` source and move common code to `source/internal`.

**Comment by mythi** [comment]
> I think for the short term the approach would be to implement new matchers to the `custom` source and move common code to `source/internal`.

I'll give it a try and if it starts to look bad, I'll get back. My only minor concern is if this results in moving code to `source/internal` and then eventually back to whatever does the discovery..

**Comment by marquiz** [comment]
Sounds like a good approach. See how it starts to shape up

**Comment by marquiz** [comment]
@mythi have you worked on this? Just interested 😉

**Comment by mythi** [comment]
@marquiz I've started moving code to `source/internal/*` today but haven't yet figured out what's the best way to deal with global configs, such as `kconfigFile`.

**Comment by marquiz** [comment]
> @marquiz I've started moving code to `source/internal/*` today but haven't yet figured out what's the best way to deal with global configs, such as `kconfigFile`

Hmm, iirc now we have something like `sources.kernel.kconfigfile`. Maybe we should introduce a new top level category like `core`, `general` or smth(?)

**Comment by marquiz** [comment]
Implemented in #334. Closing this issue

**Comment by marquiz** [comment]
Woops, IOMMU not implemented. Re-opening...

@mythi: any plans on implementing iommu rule?

**Comment by mythi** [comment]
@marquiz no plans at the moment. I can reword this issue to not include it so that we can close

**Comment by marquiz** [comment]
> @marquiz no plans at the moment. I can reword this issue to not include it so that we can close

OK, let's do that. Re-open a new issue if iommu would be needed in the future

**Comment by mythi** [comment]
cpuid and kconfig implemented in #334
